### PR TITLE
CheckpointSequential support non-reentrant

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85496
* #85495
* #86343
* #86337
* #86336
* __->__ #86331

Closes https://github.com/pytorch/pytorch/issues/86328

Adds `use_reentrant` argument to `checkpoint_sequential`.

### CODE REVIEW

**Review by zhaojuanmao — APPROVED** [review]

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
         preserve_rng_state(bool, optional):  Omit stashing and restoring
             the RNG state during each checkpoint.
             Default: ``True``
-        use_reentrant(bool, optional): Use checkpointing
+        use_reentrant(bool, optional): Use (the default) checkpointing
```
Comment (albanD) [inline torch/utils/checkpoint.py:229]: Not sure what that means? There is a `Default: True` in this section.

**Review by albanD — APPROVED** [review]
Not sure about some of the doc update. SGTM otherwise.

File: torch/utils/checkpoint.py (hunk 1 above)
Comment (albanD) [reply torch/utils/checkpoint.py:229]: I think this is already done by the `Default:` specification (note that we (try to) use this form consistently across the doc. So the user expects to find that information there already.

Comment (rohan-varma) [reply torch/utils/checkpoint.py:229]: sg, will remove

**Pull request closed without merging**